### ISSUE TITLE
Lazily create Document::fonts

### ISSUE THREAD

**Issue by jdm** [servo/servo#34037]
There's a call to `Document::Fonts()` in `Window::reflow` which can trigger a GC if it's the first time that we're calling the Fonts method. This is not ideal—if the page content never interacts with the Fonts API, we're just doing unnecessary work. To avoid this, we should:
* add a new enum in document.rs that is a boolean status value (whether the promise is fulfilled) or a `Dom<FontFaceSet>`
* make `Document::Fonts` constuct a FontFaceSet based on the value of the boolean status value if it's present, or return an existing FontFaceSet value that was previously constructured
* make `FontFaceSet::new_inherited` construct a new_resolved promise based on the passed-in status value
* update the status value from false->true in Window::reflow, or resolve the promise if document's FontFaceSet was previously constructured

This should allow us to remove the CanGc argument from Window::reflow, which should no longer be required.

**Comment by chickenleaf** [comment]
Is this an E task?

**Comment by chickenleaf** [comment]
I feel like I could do this

**Comment by jdm** [comment]
Yeah, I think it's an appropriate challenge for your skills :)

**Comment by chickenleaf** [comment]
i will need to add getter and setter methods for fonts state right? 
```
error[E0616]: field fonts_state of struct dom::document::Document is private
    --> components\script\dom\window.rs:2017:25
     |
2017 |         match *document.fonts_state.borrow() {
     |                         ^^^^^^^^^^^ private field
```
```
error[E0616]: field fonts_state of struct dom::document::Document is private
    --> components\script\dom\window.rs:2019:27
     |
2019 | ...   *document.fonts_state.borrow_mut() = Font...
     |                 ^^^^^^^^^^^ private field
```

**Comment by jdm** [comment]
Correct. Probably not getters and setters exactly, but methods that encapsulate what the reflow code actually is trying to accomplish.

**Comment by chickenleaf** [comment]
hii, i think im done with this, all that's left is removing the extra can_gc argument from reflow in rest of the codebase. i shall check the changes again before submitting the PR, is there another issue I could work on? the other two GC hazards it seems like a PR is on the way to fix them, could you please tell me if any of these would suit my current understanding of the codebase? they all seem kinda intimidating but im interested in this http api ui biznezz
https://github.com/servo/servo/issues/27439 (this one i think is similar to one of the minibrowser tasks)
https://github.com/servo/servo/issues/33662 (why has this been reopened?)
https://github.com/servo/servo/issues/31022

**Comment by jdm** [comment]
Closing this because the proposed implementation no longer avoids a GC.
